### PR TITLE
fix(dfixpnt): floor and ceil add a zeroed unit, not an uninitialized one

### LINKED ISSUE #1476: dfixpnt: floor() of a negative value subtracts an uninitialized dfixpnt (garbage, compiler-dependent)
`floor()` of a negative dfixpnt with a fractional part subtracts **uninitialized memory**, so the result is garbage and depends on the compiler.

```cpp
#include <universal/number/dfixpnt/dfixpnt.hpp>
using F = sw::universal::dfixpnt<8, 3, sw::universal::DecimalEncoding::BCD, sw::universal::Modulo, uint8_t>;
floor(F(-12.375));   // expected -13.000
```

On `main`: gcc 13.3 gives `-1.000`, clang gives `-44733.180`.

`dfixpnt/mathlib.hpp`:

```cpp
dfixpnt<...> one;            // trivially constructible: _sign and every digit are indeterminate
one.setdigit(radix, 1);      // sets ONE digit
result -= one;               // subtracts whatever else was on the stack
```

dfixpnt is deliberately trivially constructible, so a default-constructed value is not zero. The fix is to start from zero (`one.setzero();` or `dfixpnt<...> one(0);`) before `setdigit`. Check `ceil()` for the same pattern: it printed `-12.000` for `-12.375`, which is right, but worth a look.

Found by the dfixpnt layer gate (#1334). The layering PR is byte-identical and leaves it as it is.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

Fixes #1476. `floor()` of a negative value with a fraction subtracts one unit, and `ceil()` of a positive one adds it. Both built the unit as a default-constructed `dfixpnt` with one digit set. `dfixpnt` is trivially constructible, so every other digit and the sign were indeterminate. Each function was undefined behaviour on half its domain. For `dfixpnt<8, 3, BCD, Modulo, uint8_t>` at `-O2`:

| | expected | gcc 13.3 on `main` | clang on `main` |
|---|---|---|---|
| `floor(-12.375)` | -13.000 | -46513.000 | crashes inside `floor()` |
| `ceil(12.375)` | 13.000 | 46513.000 | |

The exact garbage depends on the stack, so it differs between runs and from the issue's numbers.

## Change

In `dfixpnt/mathlib.hpp`:
- **The unit** now comes from `UnitOf()`, which zeroes the value first and then sets the units digit.
- **Types with no integer digits** (`radix == ndigits`) cannot hold 1, and `setdigit(radix, 1)` wrote past the value there. With Modulo arithmetic 1 wraps to 0 anyway. So for those types the unit is 0, and `floor` and `ceil` return the integer part, 0.

A zero result keeps its operand's sign, as IEEE does: `ceil(-0.375)` prints `-0.000` and compares equal to 0. That is unchanged.

## Test

New: `static/fixpnt/decimal/math/floor_ceil.cpp`, in a new `math` folder of the dfixpnt tests wired into its `CMakeLists.txt`. It uses the standard MANUAL_TESTING and REGRESSION_LEVEL skeleton.
- **Oracle:** each value is built from a scaled integer n, and `floor` and `ceil` are checked against integer floor and ceiling division of n by 10^radix.
- **Coverage:** the BCD, BID and DPD encodings, Modulo and Saturate arithmetic, and 8- to 64-bit limbs.
- **Level 1:** exhaustive over every value of `dfixpnt<4,2>` (both modes), `<4,1>`, `<3,0>` (no fraction) and `<3,3>` (no integer digits), plus the reported cases. It samples `<8,3>`, `<12,5>` and `<16,8>`, and runs in 0.6 s.

## Verification

- **On `main`** the test segfaults at `-O2` on gcc 13.3 and clang.
- **With the fix,** levels 1-4 pass on both, and all 15 dfixpnt and blockdecimal tests pass on both. It also holds with `-ftrivial-auto-var-init=pattern`, which fills uninitialized stack memory with a fixed pattern.

## Behaviour change

`BEHAVIOUR-CHANGE: dfixpnt floor() of negative non-integers and ceil() of positive non-integers now return the right integer (they returned garbage from uninitialized memory, or crashed)`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01NazYbxaii2XXVzQeQKMDTs

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved `dfixpnt` floor and ceiling calculations for negative and positive values, including configurations with limited integer capacity.
  - Improved handling of representable fixed-point results across supported encodings and arithmetic modes.
  - Improved consistency when calculating unit values for fixed-point operations.

- **Tests**
  - Added comprehensive regression coverage for floor and ceiling operations across small and wider fixed-point configurations.
  - Added validation for edge cases involving scaling, rounding, and limited representable ranges.

### CODE REVIEW

File: static/fixpnt/decimal/math/floor_ceil.cpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+template<typename F>
+int VerifyValue(std::int64_t n, bool reportTestCases, const std::string& config) {
+	const std::int64_t unit = pow10(F::radix), limit = pow10(F::ndigits) - 1;
+	const F v = FromScaled<F>(n);
+	int fails = 0;
+	auto check = [&](const char* op, const F& got, std::int64_t expected) {
+		if (expected < -limit || expected > limit) return;  // the result does not fit the type
+		if constexpr (F::radix == F::ndigits) expected = 0;  // no integer digit: the integer part is 0
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/fixpnt/decimal/math/floor_ceil.cpp]: Fixed in c9a76ae6, and the gap was real. The no-integer-digit case is now decided before the range filter, as you proposed. For `dfixpnt<3,3>`, the reference `floor` of a negative fraction and `ceil` of a positive one is +-1000, outside `[-999, 999]`, so the exhaustive run had skipped exactly the cases that go through `UnitOf()` for such types.

I measured it with a mutant that gives `UnitOf()` a unit of 0.001 when there is no integer digit:
- the old test passes the mutant;
- the corrected test fails 5,994 cases, 1,998 per encoding (999 negative-fraction floors plus 999 positive-fraction ceils).

A second mutant, with the `radix < ndigits` guard removed so `setdigit(radix, 1)` writes past the value, gives the same results under both tests. The out-of-range digit is never read, so the guard is a bounds safeguard rather than a value change.

Levels 1-4 pass on gcc 13.3 and clang, and all 15 dfixpnt and blockdecimal tests pass on both.

**Comment by Ravenwater** [comment]
@coderabbitai review

**Comment by Ravenwater** [comment]
@coderabbitai Applied your nitpick on c9a76ae6 in b4bfa134: the project's `.clang-format` (`BreakBeforeBraces: Attach`, `BeforeElse: false`) writes `} else if (...) {` on one line, and the dfixpnt sources follow it. The test passes on gcc and clang. Nothing else changed.

**Comment by Ravenwater** [comment]
@coderabbitai review
